pymeasure/adapters/adapter.py

```diff
@@ -34,44 +37,112 @@ class Adapter:
     This class should only be inherited from.
 
     :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+        received from the instrument. The callable returns the processed
+        string. This argument is deprecated.
     :param kwargs: all other keyword arguments are ignored.
     """
 
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn(
+                "Do not use `Adapter.values`, but `Instrument.values` instead. "
+                "Preprocess_reply is a values parameter.",
+                FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
-        if self.connection is not None:
+        self.close()
+
+    def close(self):
+        """Close the connection."""
+        try:
```

Comment:
Isn't `if self.connection is not None:` more concise than a try-except?

Reply:
Maybe we have a connection, but without a close method?

Somewhere in my workings (maybe trying to test one of the instrument proper adapters) I encountered this problem.

Reply:
I'd expect such an adapter to override `Adapter.close` to do an appropriate connection closure (not silently not do it :D).

Reply:
You're right. I changed it.